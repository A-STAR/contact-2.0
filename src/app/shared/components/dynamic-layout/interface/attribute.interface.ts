import { IDynamicLayoutGenericItem, DynamicLayoutItemType } from './item.interface';

export interface IAttributeContext {
  debtId: number;
  personId: number;
  personRole: number;
  userId: number;
}

export enum AttributeType {
  FORMULA    = 'formula',
  FORMULA_ID = 'formulaId',
}

export interface IAttribute {
  type: AttributeType;
  value: string;
}

export interface IAttributeConfig {
  [key: string]: IAttribute[];
}

export interface IAttributePayload {
  attributes: IAttributeConfig;
  context: Partial<IAttributeContext>;
}

export interface IDynamicLayoutAttribute extends IDynamicLayoutGenericItem {
  attributeType: AttributeType;
  type: DynamicLayoutItemType.ATTRIBUTE;
  value: string;
  // Optional:
  dictCode?: number;
  lookupKey?: number;
}
