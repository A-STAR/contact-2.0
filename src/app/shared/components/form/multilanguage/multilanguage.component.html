<label class="m0" (click)="onLabelClick($event)">
  <div *ngIf="label" class="form-label">
    {{ label | translate }}
    <b *ngIf="isRequired" class="text-danger">*</b>
  </div>
  <div class="input-group" appDropdown>
    <span class="input-group-addon language-label" #dropdownTrigger>
      <span>
        {{ selectedLanguageLabel }}
      </span>
      <i class="caret"></i>
    </span>
    <ul class="dropdown-content" #dropdownContent>
      <li *ngFor="let language of languages" (click)="onLanguageSelect(language)">
        {{ language.name }}
      </li>
    </ul>
    <input
      #input
      class="form-control"
      type="text"
      [disabled]="isDisabled"
      [ngModel]="selectedTranslationLabel"
      (focusout)="onFocusOut()"
      (input)="onChange($event)"
    />
  </div>
  <div>
    <span *ngIf="errors">{{ errors | json }}</span>
  </div>
</label>
