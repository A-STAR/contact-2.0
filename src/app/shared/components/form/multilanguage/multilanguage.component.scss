.dropdown-content {
  background-color: #fff;
  list-style-type: none;
  margin: 0;
  padding: 5px 0;
  z-index: 20000;

  > li {
    cursor: pointer;
    padding: 5px 10px;

    &:hover {
      background-color: #eee;
    }
  }
}

.language-label {
  display: inline-block;
  text-align: left;
  width: 150px;
}

.input-group-addon {
  cursor: pointer;
}
