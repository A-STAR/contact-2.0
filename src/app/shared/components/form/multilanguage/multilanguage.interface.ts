export interface IMultiLanguageConfig {
  entityAttributeId: number;
  entityId: number;
}
