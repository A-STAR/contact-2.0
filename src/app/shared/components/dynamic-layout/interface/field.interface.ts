import { IDynamicLayoutGenericItem, DynamicLayoutItemType } from './item.interface';

export interface IDynamicLayoutField extends IDynamicLayoutGenericItem {
  type: DynamicLayoutItemType.FIELD;
}
