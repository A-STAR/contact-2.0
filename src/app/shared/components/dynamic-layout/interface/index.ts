import { Observable } from 'rxjs/Observable';

import { IDynamicLayoutControl } from './control.interface';
import { IDynamicLayoutGroup } from './group.interface';
import { IDynamicLayoutTemplate } from './template.interface';

export * from './context.interface';
export * from './control.interface';
export * from './attribute.interface';
export * from './group.interface';
export * from './item.interface';
export * from './template.interface';

export type IDynamicLayoutItem = IDynamicLayoutGroup | IDynamicLayoutControl | IDynamicLayoutTemplate;

export interface IDynamicLayoutConfig {
  items: IDynamicLayoutItem[];
}

export interface IDynamicLayoutItemProperties {
  item: IDynamicLayoutItem;
  streams: {
    display: Observable<boolean>;
  };
}
