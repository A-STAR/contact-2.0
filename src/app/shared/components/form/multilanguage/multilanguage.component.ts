import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ElementRef,
  forwardRef,
  Input,
  OnInit,
  ViewChild,
} from '@angular/core';
import { ControlValueAccessor, NG_VALIDATORS, NG_VALUE_ACCESSOR, Validator } from '@angular/forms';
import { combineLatest } from 'rxjs/observable/combineLatest';
import { of } from 'rxjs/observable/of';

import { IEntityTranslation } from '@app/core/entity/translations/entity-translations.interface';
import { ILookupLanguage } from '@app/core/lookup/lookup.interface';
import { IMultiLanguageConfig } from '@app/shared/components/form/multilanguage/multilanguage.interface';

import { DataService } from '@app/core/data/data.service';
import { LookupService } from '@app/core/lookup/lookup.service';

import { DropdownDirective } from '@app/shared/components/dropdown/dropdown.directive';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => MultiLanguageComponent),
      multi: true,
    },
    {
      provide: NG_VALIDATORS,
      useExisting: forwardRef(() => MultiLanguageComponent),
      multi: true,
    },
  ],
  selector: 'app-multilanguage',
  styleUrls: [ './multilanguage.component.scss' ],
  templateUrl: './multilanguage.component.html',
})
export class MultiLanguageComponent implements ControlValueAccessor, OnInit, Validator {
  @Input() errors: any;
  @Input() isDisabled = false;
  @Input() isRequired = false;
  @Input() label: string;
  @Input() langConfig: IMultiLanguageConfig;

  @ViewChild('input') input: ElementRef;
  @ViewChild(DropdownDirective) dropdown: DropdownDirective;

  languages: ILookupLanguage[] = [];
  value: IEntityTranslation[] = [];

  private mainLanguageId: number;
  private selectedLanguageId = 1;

  constructor(
    private cdRef: ChangeDetectorRef,
    private dataService: DataService,
    private lookupService: LookupService,
  ) {}

  get selectedLanguageLabel(): string {
    const language = this.languages.find(l => l.id === this.selectedLanguageId);
    return language ? language.name : null;
  }

  get selectedTranslationLabel(): string {
    const translation = this.value.find(v => v.languageId === this.selectedLanguageId);
    return translation ? translation.value : null;
  }

  ngOnInit(): void {
    const { entityId, entityAttributeId } = this.langConfig;

    combineLatest(
      this.lookupService.lookup<ILookupLanguage>('languages'),
      // This is where multilanguage controls get different from other controls.
      // We have to get value not from form but from server api.
      entityId && entityAttributeId
        ? this.dataService.readTranslations(entityId, entityAttributeId)
        : of(null),
    ).subscribe(([ languages, value ]) => {
      this.languages = languages;
      this.value = value || languages.map(l => ({ languageId: l.id, value: null }));

      const mainLanguage = languages.find(l => l.isMain === 1);
      if (mainLanguage) {
        this.mainLanguageId = mainLanguage.id;
      }

      this.propagateChange(this.value);
      this.cdRef.markForCheck();
    });
  }

  writeValue(value: any[]): void {
    if (Array.isArray(value)) {
      this.value = value;
      this.cdRef.markForCheck();
    }
  }

  registerOnChange(fn: Function): void {
    this.propagateChange = fn;
  }

  registerOnTouched(fn: Function): void {
    this.propagateTouch = fn;
  }

  setDisabledState(disabled: boolean): void {
    this.isDisabled = disabled;
  }

  validate(): any {
    const mainItem = this.mainLanguageId
      ? this.value.find(i => i.languageId === this.mainLanguageId)
      : null;

    switch (true) {
      case this.isRequired && mainItem && !mainItem.value:
        return { required: true };
      default:
        return null;
    }
  }

  onChange(event: Event): void {
    const { value } = event.target as HTMLInputElement;
    const item = this.value.find(i => i.languageId === this.selectedLanguageId);
    if (item) {
      item.value = value || null;
    } else {
      this.value.push({
        languageId: this.selectedLanguageId,
        value: value || null,
      });
    }
    this.propagateChange(this.value);
  }

  onFocusOut(): void {
    this.propagateTouch();
  }

  onLanguageSelect(language: any): void {
    this.selectedLanguageId = language.id;
    this.dropdown.close();
    this.input.nativeElement.focus();
    this.cdRef.markForCheck();
  }

  onLabelClick(event: MouseEvent): void {
    event.preventDefault();
    this.input.nativeElement.focus();
  }

  private propagateChange: Function = () => {};
  private propagateTouch: Function = () => {};
}
