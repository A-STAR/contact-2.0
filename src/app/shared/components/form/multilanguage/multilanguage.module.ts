import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';

import { DropdownModule } from '@app/shared/components/dropdown/dropdown.module';

import { MultiLanguageComponent } from './multilanguage.component';

@NgModule({
  imports: [
    CommonModule,
    DropdownModule,
    FormsModule,
    TranslateModule,
  ],
  exports: [
    MultiLanguageComponent,
  ],
  declarations: [
    MultiLanguageComponent,
  ],
})
export class MultiLanguageModule {}
