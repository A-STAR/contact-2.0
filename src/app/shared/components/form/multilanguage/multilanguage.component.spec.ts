import { async as Async, ComponentFixture, TestBed } from '@angular/core/testing';
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { FormsModule } from '@angular/forms';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import { IEntityTranslation } from '@app/core/entity/translations/entity-translations.interface';
import { ILookupLanguage } from '@app/core/lookup/lookup.interface';

import { DataService } from '@app/core/data/data.service';
import { LookupService } from '@app/core/lookup/lookup.service';

import { MultiLanguageComponent } from './multilanguage.component';

class TranslateLoaderMock {
  getTranslation(language: string): Observable<any> {
    return of([]);
  }
}

class DataServiceMock {
  readTranslations(entityId: number, entityAttributeId: number): Observable<IEntityTranslation[]> {
    return of([
      { languageId: 1, value: 'Value for language 1' },
      { languageId: 2, value: 'Value for language 2' },
    ]);
  }
}

class LookupServiceMock {
  lookup(key: string): Observable<ILookupLanguage[]> {
    return of([
      { id: 1, name: 'Language 1', isMain: 1 },
      { id: 2, name: 'Language 2', isMain: 0 },
    ]);
  }
}

describe('MultiLanguageComponent', () => {
  let fixture: ComponentFixture<MultiLanguageComponent>;

  beforeEach(Async(() => {
    TestBed
      .configureTestingModule({
        declarations: [
          MultiLanguageComponent,
        ],
        imports: [
          FormsModule,
          TranslateModule.forRoot({
            loader: {
              provide: TranslateLoader,
              useClass: TranslateLoaderMock,
            },
          }),
        ],
        providers: [
          {
            provide: DataService,
            useClass: DataServiceMock,
          },
          {
            provide: LookupService,
            useClass: LookupServiceMock,
          },
        ],
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(MultiLanguageComponent);
  });

  it('should render', () => {
    fixture.componentInstance.langConfig = {
      entityId: 1,
      entityAttributeId: 1,
    };
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
