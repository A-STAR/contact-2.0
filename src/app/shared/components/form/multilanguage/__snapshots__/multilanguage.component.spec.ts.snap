// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiLanguageComponent should render 1`] = `
<div
  id="root0"
  ng-version="5.2.4"
>
  <label
    class="m0"
  >
    
  
    
    
  
    <div
      appdropdown=""
      class="input-group"
    >
      
    
      <span
        class="input-group-addon language-label"
      >
        
      
        <span>
          
        Language 1
      
        </span>
        
      
        <i
          class="caret"
        />
        
    
      </span>
      
    
      <ul
        class="dropdown-content"
      >
        
      
        
        <li>
          
        Language 1
      
        </li>
        <li>
          
        Language 2
      
        </li>
        
    
      </ul>
      
    
      <input
        class="form-control ng-untouched ng-pristine ng-valid"
        ng-reflect-is-disabled="false"
        ng-reflect-model="Value for language 1"
        type="text"
      />
      
  
    </div>
    
  
    <div>
      
    
      
      
  
    </div>
    

  </label>
  

</div>
`;
